### PR TITLE
Add vulnerability location to protobuf and agent_report_vulnerability_mixin.

### PR DESCRIPTION
**Pull request by mohsinenar**

### CODE REVIEW

File: src/ostorlab/agent/mixins/agent_report_vulnerability_mixin.py [hunk 1]
```diff
@@ -3,11 +3,55 @@
...
+@dataclasses.dataclass
+class AssetLocation:
+    """Asset location used to attach a vulnerability to a specific asset"""
+    metadata: List[AssetLocationMetaData]
+    asset: asset.Asset
+
+    def to_dict(self) -> dict:
+        """Convert data class to a dict matching what is expected from protobuf."""
+        return {
+            MAP_ASSETS_NAMES[self.asset.__class__.__name__]: self.asset.__dict__,
```
Comment (deadly-panda) [inline src/ostorlab/agent/mixins/agent_report_vulnerability_mixin.py:52]: I would've put the classes as keys of MAP_ASSETS_NAMES, & use type(self.asset) here.

Comment (deadly-panda) [inline src/ostorlab/agent/mixins/agent_report_vulnerability_mixin.py:52]: Also I would've used: `dataclasses.asdict(self.asset)` and avoid any magic functions.

File: src/ostorlab/agent/mixins/agent_report_vulnerability_mixin.py [hunk 2]
```diff
@@ -29,9 +74,11 @@ def report_vulnerability(self,
                              entry: kb.Entry,
                              technical_detail: str,
                              risk_rating: RiskRating,
-                             dna: Optional[str] = None) -> None:
+                             dna: Optional[str] = None,
+                             asset_location: AssetLocation = None) -> None:
```
Comment (deadly-panda) [inline src/ostorlab/agent/mixins/agent_report_vulnerability_mixin.py:78]: The name `asset_location` is confusing.

File: src/ostorlab/agent/mixins/agent_report_vulnerability_mixin.py [hunk 3]
```diff
@@ -73,7 +119,8 @@ def report_vulnerability(self,
             'targeted_by_ransomware': targeted_by_ransomware,
             'targeted_by_nation_state': targeted_by_nation_state,
             'cvss_v3_vector': cvss_v3_vector,
-            'dna': dna
+            'dna': dna,
+            'asset_location': asset_location.to_dict() if asset_location is not None else None
```
Comment (deadly-panda) [inline src/ostorlab/agent/mixins/agent_report_vulnerability_mixin.py:123]: If it is None, do not emit it.

File: src/ostorlab/assets/ios_store.py [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+@dataclasses.dataclass
+@asset.selector('v3.asset.store.ios_store')
+class IOSStore(asset.Asset):
+    """Ios bundle target asset."""
+
+    def __init__(self, bundle_id: str):
+        self.bundle_id = bundle_id
+
+    def __str__(self):
+        return f'IOS store bundle_id ({self.path})'
```
Comment (deadly-panda) [inline src/ostorlab/assets/ios_store.py:16]:
```suggestion
        return f'IOS store bundle_id ({self.bundle_id})'
```

File: tests/agent/mixins/agent_report_vulnerability_mixin_test.py [hunk 5]
```diff
@@ -31,3 +33,44 @@ def testSendVulnerabilityMessage_whenKbEntryIsValid_emitVulnerabilityMessage(age
...
+                          'asset_location_android_apk'])
+def testSendVulnerabilityMessageWithAssetLocation_whenKbEntryIsValid_emitVulnerabilityMessage(asset_location_fixture,
+                                                                                              request,
+                                                                                              agent_mock: List[object],
+                                                                                              ) -> None:
+    """Prepares & emit vulnerability message with all the information from the Knowledge base."""
+    agent_definition = agent_definitions.AgentDefinition(name='some_name', out_selectors=['v3.report.vulnerability'])
+    agent_settings = runtime_definitions.AgentSettings(key='some_key')
+    report_vul_mixin = TestAgent(agent_definition, agent_settings)
+    request.getfixturevalue(asset_location_fixture)
```
Comment (deadly-panda) [inline tests/agent/mixins/agent_report_vulnerability_mixin_test.py:58]: is this needed here ?

File: tests/conftest.py [hunk 6]
```diff
@@ -225,3 +236,101 @@ def clean_redis_data(request) -> None:
...
+    return agent_report_vulnerability_mixin.AssetLocation(
+        metadata=[metadata_file_path, metadata_code_location, metadata_port, metadata_url],
+        asset=ipv4_asset.IPv4(host='8.8.8.8'))
+
+
+@pytest.fixture()
+def asset_location_ipv6(metadata_file_path, metadata_code_location, metadata_port, metadata_url):
+    return agent_report_vulnerability_mixin.AssetLocation(
+        metadata=[metadata_file_path, metadata_code_location, metadata_port, metadata_url],
+        asset=ipv6_asset.IPv6(host='8.8.8.8'))
```
Comment (deadly-panda) [inline tests/conftest.py:322]:
```suggestion
        asset=ipv6_asset.IPv6(host='2001:4860:4860::8888'))
```

File: src/ostorlab/agent/mixins/agent_report_vulnerability_mixin.py [hunk 7]
```diff
@@ -29,9 +84,11 @@ def report_vulnerability(self,
                              entry: kb.Entry,
                              technical_detail: str,
                              risk_rating: RiskRating,
-                             dna: Optional[str] = None) -> None:
+                             dna: Optional[str] = None,
+                             vulnerability_location: VulnerabilityLocation = None) -> None:
```
Comment (deadly-panda) [inline src/ostorlab/agent/mixins/agent_report_vulnerability_mixin.py:88]:
```suggestion
                             vulnerability_location: Optional[VulnerabilityLocation] = None) -> None:
```

File: src/ostorlab/agent/mixins/agent_report_vulnerability_mixin.py [hunk 8]
```diff
@@ -73,7 +129,8 @@ def report_vulnerability(self,
             'targeted_by_ransomware': targeted_by_ransomware,
             'targeted_by_nation_state': targeted_by_nation_state,
             'cvss_v3_vector': cvss_v3_vector,
-            'dna': dna
+            'dna': dna,
+            'vulnerability_location': vulnerability_location.to_dict() if vulnerability_location is not None else None
```
Comment (deadly-panda) [inline src/ostorlab/agent/mixins/agent_report_vulnerability_mixin.py:133]: don't put it in the dict if it's None.

**Review by deadly-panda — CHANGES_REQUESTED** [review]

File: src/ostorlab/agent/message/proto/v3/report/vulnerability/vulnerability.proto [hunk 9]
```diff
@@ -30,17 +34,21 @@ enum MetaDataTypeEnum {
...
-message AssetLocation {
+message VulnerabilityLocation {
   oneof asset {
     ostorlab.agent.message.proto.v3.asset.domain_name.Message domain_name = 1;
     ostorlab.agent.message.proto.v3.asset.ip.v4.Message ipv4 = 2;
     ostorlab.agent.message.proto.v3.asset.ip.v6.Message ipv6 = 3;
     ostorlab.agent.message.proto.v3.asset.link.Message link = 4;
     ostorlab.agent.message.proto.v3.asset.store.android_store.Message android_store = 5;
     ostorlab.agent.message.proto.v3.asset.store.ios_store.Message ios_store = 6;
+    ostorlab.agent.message.proto.v3.asset.file.Message  file = 7;
```
Comment (3asm) [inline src/ostorlab/agent/message/proto/v3/report/vulnerability/vulnerability.proto:48]:
```suggestion
    ostorlab.agent.message.proto.v3.asset.file.Message file = 7;
```

File: src/ostorlab/agent/message/proto/v3/report/vulnerability/vulnerability.proto [hunk 10]
```diff
@@ -30,17 +34,21 @@ enum MetaDataTypeEnum {
...
+    ostorlab.agent.message.proto.v3.asset.file.android.aab.Message  android_aab = 8;
+    ostorlab.agent.message.proto.v3.asset.file.android.apk.Message  android_apk = 9;
```
Comment (3asm) [inline src/ostorlab/agent/message/proto/v3/report/vulnerability/vulnerability.proto:50]:
```suggestion
    ostorlab.agent.message.proto.v3.asset.file.android.aab.Message android_aab = 8;
    ostorlab.agent.message.proto.v3.asset.file.android.apk.Message android_apk = 9;
```

File: src/ostorlab/agent/mixins/agent_report_vulnerability_mixin.py [hunk 11]
```diff
@@ -3,11 +3,65 @@
...
+    android_aab_asset.AndroidAab: 'android_aab',
+    android_apk_asset.AndroidApk: 'android_apk',
+    ios_ipa_asset.IOSIpa: 'ios_ipa',
+    android_store_asset.AndroidStore: 'android_store',
+    ios_store_asset.IOSStore: 'ios_store',
+}
+
+
+class MetaDataType(enum.Enum):
+    """Asset Location MetaData Type"""
```
Comment (3asm) [inline src/ostorlab/agent/mixins/agent_report_vulnerability_mixin.py:39]:
```suggestion
    """Asset location metadata type."""
```

File: src/ostorlab/agent/mixins/agent_report_vulnerability_mixin.py [hunk 12]
```diff
@@ -3,11 +3,65 @@
...
+from ostorlab.assets import domain_name as domain_name_asset
+from ostorlab.assets import file as file_asset
+from ostorlab.assets import ios_ipa as ios_ipa_asset
+from ostorlab.assets import ios_store as ios_store_asset
+from ostorlab.assets import ipv4 as ipv4_asset
+from ostorlab.assets import ipv6 as ipv6_asset
+from ostorlab.assets import link as link_asset
+
+MAP_ASSETS_NAMES = {
+    ipv4_asset.IPv4: 'ipv4',
```
Comment (3asm) [inline src/ostorlab/agent/mixins/agent_report_vulnerability_mixin.py:25]: The string thingy should an attribute of each asset. Centralizing it here is change error prone.

File: src/ostorlab/agent/mixins/agent_report_vulnerability_mixin.py [hunk 13]
```diff
@@ -3,11 +3,65 @@
...
+    FILE_PATH = enum.auto()
+    CODE_LOCATION = enum.auto()
+    URL = enum.auto()
+    PORT = enum.auto()
+
+
+@dataclasses.dataclass
+class VulnerabilityLocationMetaData:
+    """Asset location metadata holding the type and the value of 'where' the vulnerability was detected in the asset """
+    type: MetaDataType
```
Comment (3asm) [inline src/ostorlab/agent/mixins/agent_report_vulnerability_mixin.py:49]: Metadata not MetaData. Please change everywhere.

File: src/ostorlab/agent/mixins/agent_report_vulnerability_mixin.py [hunk 14]
```diff
@@ -3,11 +3,65 @@
...
+    value: str
+
+
+@dataclasses.dataclass
+class VulnerabilityLocation:
+    """Asset location used to attach a vulnerability to a specific asset"""
+    metadata: List[VulnerabilityLocationMetaData]
+    asset: asset.Asset
+
+    def to_dict(self) -> dict:
```
Comment (3asm) [inline src/ostorlab/agent/mixins/agent_report_vulnerability_mixin.py:59]: dict of what? Please ensure typing is correct and also run mypy on it this part.

File: src/ostorlab/agent/mixins/agent_report_vulnerability_mixin.py [hunk 15]
```diff
@@ -29,9 +84,11 @@ def report_vulnerability(self,
                              entry: kb.Entry,
                              technical_detail: str,
                              risk_rating: RiskRating,
-                             dna: Optional[str] = None) -> None:
+                             dna: Optional[str] = None,
+                             vulnerability_location: Optional[VulnerabilityLocation] = None) -> None:
         """Fetch the details of an entry from the knowledge base, and emit a vulnerability message.
         Args:
+            vulnerability_location:  asset location where the vulnerability was detected.
```
Comment (3asm) [inline src/ostorlab/agent/mixins/agent_report_vulnerability_mixin.py:91]: Please respect the order of the arguments.

File: src/ostorlab/assets/android_store.py [hunk 16]
```diff
@@ -0,0 +1,16 @@
...
+@dataclasses.dataclass
+@asset.selector('v3.asset.store.android_store')
+class AndroidStore(asset.Asset):
+    """Android store package target asset."""
+
+    def __init__(self, package_name: str):
+        self.package_name = package_name
+
+    def __str__(self):
+        return f'Android Store: {self.package_name}'
```
Comment (3asm) [inline src/ostorlab/assets/android_store.py:16]: Please aligh the string to have the same format. Android Soter ({self.package_name}) is coherent with the rest.

File: src/ostorlab/assets/ios_store.py [hunk 17]
```diff
@@ -0,0 +1,16 @@
...
+@dataclasses.dataclass
+@asset.selector('v3.asset.store.ios_store')
+class IOSStore(asset.Asset):
+    """Ios bundle target asset."""
+
+    def __init__(self, bundle_id: str):
+        self.bundle_id = bundle_id
+
+    def __str__(self):
+        return f'IOS store bundle_id ({self.bundle_id})'
```
Comment (3asm) [inline src/ostorlab/assets/ios_store.py:16]:
```suggestion
        return f'iOS Store ({self.bundle_id})'
```

Coherence, coherence, coherence!

File: tests/agent/mixins/agent_report_vulnerability_mixin_test.py [hunk 18]
```diff
@@ -31,3 +33,45 @@ def testSendVulnerabilityMessage_whenKbEntryIsValid_emitVulnerabilityMessage(age
...
+                          ('vulnerability_location_ios_store', 'ios_store'),
+                          ('vulnerability_location_ipv4', 'ipv4'),
+                          ('vulnerability_location_ipv6', 'ipv6'),
+                          ('vulnerability_location_domain_name', 'domain_name'),
+                          ('vulnerability_location_android_store', 'android_store'),
+                          ('vulnerability_location_ios_ipa', 'ios_ipa'),
+                          ('vulnerability_location_link_asset', 'link'),
+                          ('vulnerability_location_file', 'file'),
+                          ('vulnerability_location_android_apk', 'android_apk')])
+def testSendVulnerabilityMessageWithAssetLocation_whenKbEntryIsValid_emitVulnerabilityMessage(
```
Comment (3asm) [inline tests/agent/mixins/agent_report_vulnerability_mixin_test.py:50]:
```suggestion
def testSendVulnerabilityMessage_withVulnerabilityLocation_emitVulnerabilityMessageWithProperData(
```

File: tests/assets/asset_test.py [hunk 19]
```diff
@@ -6,11 +6,11 @@
 
 
 def testAssetToProto_whenSelectorIsSetAndCorrect_generatesProto():
-
     @dataclasses.dataclass
     @asset.selector('v3.asset.file.android.apk')
     class SimpleAndroidApk(asset.Asset):
-        content: bytes
+        def __init__(self, content):
```
Comment (3asm) [inline tests/assets/asset_test.py:12]: It is a dataclass, why do you need to change the init method?

**Review by 3asm — CHANGES_REQUESTED** [review]

**Review by 3asm — APPROVED** [review]

**Pull request merged**